src/core/cache.ts

```diff
@@ -96,19 +96,30 @@ export const computeTotalSize = (cache: Cache): number => {
  * @internal
  */
 export const findIndex = (cache: Cache, offset: number, i: number): number => {
-  while (i >= 0 && i < cache._length) {
-    const itemOffset = computeOffset(cache, i);
+  let low = 0;
+  let high = cache._length - 1;
+  let mid, itemOffset;
+
+  if (computeOffset(cache, i) <= offset) {
+    low = i; // Start searching from initialIndex -> up
+  } else {
+    high = i; // Start searching from initialIndex -> down
+  }
+
+  while (low <= high) {
+    mid = Math.floor((low + high) / 2);
+    itemOffset = computeOffset(cache, mid);
+
     if (itemOffset <= offset) {
-      if (itemOffset + getItemSize(cache, i) > offset) {
-        break;
-      } else {
-        i++;
+      if (itemOffset + getItemSize(cache, mid) > offset) {
+        return clamp(mid, 0, cache._length - 1);
       }
+      low = mid + 1;
     } else {
-      i--;
+      high = mid - 1;
     }
   }
-  return clamp(i, 0, cache._length - 1);
+  return clamp(low, 0, cache._length - 1);
```

Comment:
Does returning with `low` make sense here? @inokawa

Reply:
Do you mean this path is unreachable? If so we should just throw error. In the past, there were some cases the index exceeded the limit, but now I'm not sure.

Reply:
I think it may be possible to not find an index or have invalid input args; usually I lean toward early exit with `-1` on invalid input args and a default return of `-1` for binary search stuff. FWIW though, there isn't anything in place at the moment in the consumers of `findIndex` to handle this behavior. 

I'm happy to discuss approaches or implement your preferred approach to that issue in this PR, but ATM I think this should mimic the current behavior with just a change in how the index is searched for.

Reply:
In the previous implementation, though logically it may not be possible given the constraints on input args, it could exit with `i=0` or `i=cache._length`, both of which are not a "found index" technically so it may lead to undesirable behavior.

If we were to return `-1` when no index is found we could check for this in the consumer and prevent subsequent behavior.